self-api/src/main/java/com/selfxdsd/api/ConditionalResource.java

```diff
@@ -0,0 +1,213 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
+ * OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
+ * OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.selfxdsd.api;
+
+import javax.json.JsonArray;
+import javax.json.JsonObject;
+import java.net.URI;
+import java.time.LocalDateTime;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Conditional {@link Resource}.
+ * A Resource is considered conditional if it has a eTag header in its
+ * response headers.
+ * @author criske
+ * @version $Id$
+ * @since 0.0.8
+ */
+public interface ConditionalResource extends Resource {
+
+    /**
+     * Uri of this resource. This should be also considered the
+     * storing key.
+     *
+     * @return URI.
+     */
+    URI uri();
+
+    /**
+     * ETag of this resource.
+     *
+     * @return String.
+     */
+    String etag();
+
+    /**
+     * Creation date of the conditional resource.
+     *
+     * @return LocalDateTime.
+     */
+    LocalDateTime creationDate();
+
+    /**
+     * Factory method that creates a ConditionalResource from a regular
+     * Resource.
+     * If the regular can't be conditional
+     * (no ETag header present in its response headers) it returns null.
+     *
+     * @param uri URI associated with the resource.
+     * @param resource Original Resource.
+     * @return ConditionalResource or null.
+     */
+    static ConditionalResource fromResource(
+        final URI uri,
+        final Resource resource
+    ) {
+        final ConditionalResource conditional;
+        if (resource instanceof ConditionalResource) {
+            conditional = (ConditionalResource) resource;
+        } else {
+            final String eTag = findEtag(resource);
+            if (eTag != null) {
+                conditional = new FromResource(resource, eTag, uri);
+            } else {
+                conditional = null;
+            }
+        }
+        return conditional;
+    }
+
+    /**
+     * Get ETag from resource headers.
+     *
+     * @param resource Resource.
+     * @return String or null if ETag was not found.
+     */
+    private static String findEtag(final Resource resource) {
+        return resource
+            .headers()
+            .entrySet()
+            .stream()
+            .filter(entry -> entry.getKey()
+                .equalsIgnoreCase("ETag"))
+            .flatMap(entry -> entry.getValue().stream())
+            .findFirst()
+            .orElse(null);
```

Comment:
@criske this should be inside FromHeaders.eTag()

Reply:
@amihaiemil I agree, this could be inlined... but etag is already provided with CachedResource constructor. Again, the factory handles this.